Source/engine/render/text_render.cpp

```diff
@@ -396,17 +397,23 @@ int GetLineStartX(UiFlags flags, const Rectangle &rect, int lineWidth)
 }
 
 uint32_t DoDrawString(const Surface &out, std::string_view text, Rectangle rect, Point &characterPosition,
-    int lineWidth, int rightMargin, int bottomMargin, GameFontTables size, text_color color, bool outline,
-    TextRenderOptions &opts)
+	int lineWidth, int rightMargin, int bottomMargin, GameFontTables size, text_color color, bool outline,
+	TextRenderOptions &opts)
 {
 	CurrentFont currentFont;
 
-	char32_t next;
-	std::string_view remaining = text;
-	size_t cpLen;
+	std::u32string text32 = ConvertUtf8ToUtf32(text);

```

Comment:
We like to avoid allocations in code that runs on every frame, is it possible to implement this in a streaming fashion?
For example, by finding the spans that contain Hebrew text and decoding and drawing them from the end?
I can add a `DecodeLastUtf8CodePoint` function if it helps.

It looks like fribidi only supports UTF-32 but perhaps there is another library?
A quick search reveals https://github.com/Tehreer/SheenBidi, which supports UTF-8.

Reply:
Thank you, 2 notes regarding this (from my point of view, please feel free to suggest otherwise)
1. Because I am trying to find and handle each line separately, working with UTF-32 actually greatly simplify the indexing of the string, allowing simpler logic to get each line, which I find harder doing on UTF-8.
2. I was having trouble setting up additional dependency as I am not yet experienced with cmake. I was able to set up linking with `freebidi` binary on linux because it is quite popular. I am open to alternative but I would need guidance for setting them up.

Reply:
> Because I am trying to find and handle each line separately, working with UTF-32 actually greatly simplify the indexing of the string, allowing simpler logic to get each line, which I find harder doing on UTF-8.

UTF-8 is a self-synchronizing encoding, which means line breaks can be found simply via `find('\n')`, without the need for decoding.

> I am open to alternative but I would need guidance for setting them up.

It's quite simple, add a file `3rdParty/SheenBidi/CMakeLists.txt` with:

```cmake
include(functions/FetchContent_MakeAvailableExcludeFromAll)

set(SOL2_ENABLE_INSTALL OFF)

include(FetchContent)
FetchContent_Declare(SheenBidi
    URL https://github.com/glebm/SheenBidi/archive/3721411b5ec2862240f34aeeb3e8ba59283ec2ec.tar.gz
    URL_HASH MD5=7dc0138bda1e16b217cee66cfd95a6f9
)
FetchContent_MakeAvailableExcludeFromAll(SheenBidi)
```

The target to link to is `SheenBidi::sheenbidi`.